### ISSUE TITLE
Uninstalling a mod should also prompt user about uninstalling its unused dependencies

### ISSUE THREAD

**Issue by Dazpoet** [KSP-CKAN/CKAN#978]
It was mentioned in the CKAN thread over at the forums that it would be pretty kickass if CKAN upon removing a mod also removed dependencies which no other mod depends on.

Say we e.g. install `FerramAerospaceResearch` which depends on `ModuleManager` and `ModularFlightIntegrator`. After that we install `HotRockets` which depends on `ModuleManager` and `SmokeScreen`.

We now have the following list of installed mods:

```
- FerramAerospaceResearch
- ModuleManager
- ModularFlightIntegrator
- HotRockets
- SmokeScreen
```

Currently if we uninstall `FerramAerospaceResearch` we'll end up with the following list of installed mods: 

```
- ModuleManager
- ModularFlightIntegrator
- HotRockets
- SmokeScreen
```

While it seems more logical that we should end up with:

```
- ModuleManager
- HotRockets
- SmokeScreen
```

The reason for prompting the user about these things would be that e.g. uninstalling `KarbonitePlus` doesn't necessarily mean I want to uninstall `Karbonite`.

I think e.g. apt already works this way so it would seem a logical direction for CKAN to go in.

**Comment by dbent** [comment]
Arch Linux's package manager, pacman, (and presumably others) does this by keeping a bit in its database to indicate if a package was installed explicitly by the user. Then when a package is uninstalled, it will also automatically remove any dependencies that were 1) not installed explicitly and 2) are not needed by any other package.

pacman also includes functionality to:

- Flip the explicit bit on/off
- Control whether or not automatic removal  is used
- Perform the removal recursively (so if Mod A depends on Mod B which depends on Mod C, removing Mod A will automatically remove Mod B *and* Mod C, so long as neither were installed explicitly and nothing depends on them)
- List "orphan" packages, which are packages with the explicit bit off but are currently not used by any other package.

By default I would suggest that automatic removal be used *and* be recursive. I would also suggest that since "explicitness" isn't currently stored that when first populating the bit, all top level packages (packages that are not dependencies for any other package) be marked as explicitly installed and all other packages be marked as being installed implicitly. This isn't perfect but *good enough* so long as users can toggle the explicit bit themselves.

**Comment by pjf** [comment]
+1 on dbent's comments here.

**Comment by Dazpoet** [comment]
I like any kind of solution that makes this system work, think it would be awesome :)

**Comment by Dazpoet** [comment]
Implementing this might accidentally provide a solution to https://github.com/KSP-CKAN/CKAN/issues/589 ? Or am I misunderstanding how this marking of mods explicitly installed and not would be added?

**Comment by pjf** [comment]
@Dazpoet : Aye, this is essentially #589 with an implementation strategy. :)

**Comment by HebaruSan** [comment]
Implemented in #2753.
